azure-batch/azure/batch/custom/patch.py

```diff
@@ -0,0 +1,285 @@
+
+        # Variables to be used for task add_collection requests
+        self._client = TaskOperations(
+            client._client, client.config, client._serialize, client._deserialize)
+        self._job_id = job_id
+        self._task_add_collection_options = task_add_collection_options
+        self._custom_headers = custom_headers
+        self._raw = raw
+        self._kwargs = dict(**kwargs)
+
+    def _bulk_add_tasks(self, results_queue, chunk_tasks_to_add):
+        """Adds a chunk of tasks to the job and retry if chunk
+
+        :param results_queue: Queue to place the return value of the request
+        :type results_queue: collections.deque
+        :param chunk_tasks_to_add: Chunk of at most 100 tasks with retry details
+        :type chunk_tasks_to_add: list[~TrackedCloudTask]
+        """
+
+        try:
+            add_collection_response = self._client.add_collection(
+                self._job_id,
+                chunk_tasks_to_add,
+                self._task_add_collection_options,
+                self._custom_headers,
+                self._raw)
+        except BatchErrorException as e:
+            # In case of a chunk exceeding the MaxMessageSize split chunk in half
+            # and resubmit smaller chunk requests
+            # TODO: Replace string with constant variable once available in SDK
+            if e.error.code == "RequestBodyTooLarge":  # pylint: disable=no-member
+                # In this case the task is misbehaved and will not be able to be added due to:
+                #   1) The task exceeding the max message size
+                #   2) A single cell of the task exceeds the per-cell limit, or
+                #   3) Sum of all cells exceeds max row limit
+                if len(chunk_tasks_to_add) == 1:
+                    failed_task = chunk_tasks_to_add.pop()
+                    results_queue.appendleft(e)
+                    logging.error("Task ID " + failed_task.id + " failed to add due to exceeding the request body" +
+                                    " being too large")
+                    self._has_early_termination_error = True
+                else:
+                    # Assumption: Tasks are relatively close in size therefore if one batch exceeds size limit
+                    # we should decrease the initial task collection size to avoid repeating the error
+                    # Midpoint is lower bounded by 1 due to above base case
+                    midpoint = int(len(chunk_tasks_to_add) / 2)
+                    # Restrict one thread at a time to do this compare and set,
+                    # therefore forcing max_tasks_per_request to be strictly decreasing
+                    with self._max_tasks_lock:
+                        if midpoint < self._max_tasks_per_request:
+                            self._max_tasks_per_request = midpoint
+                            logging.info("Amount of tasks per request reduced from " +
```

Comment:
Never log with "+" concat. Use the comma syntax, otherwise your evaluate your string even if the log level is undefined:
https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html